### DISCUSSION TITLE
Why does running Alire's Printenv in GNAT Studio report an error?

### DISCUSSION THREAD

**Discussion by Joebeazelman** [alire-project/alire#1276] [Q&A]
Whenever I run  **alr printenv** from the Gnat Studio's OS terminal window to look at the environment variables, I get the following error:

> "ERROR: Trying to set an already defined environment variable:\nERROR:    HAL_ALIRE_PREFIX is already defined as '/Users/Joe/Projects/Repos/ADBController/alire/cache/dependencies/hal_0.3.0_095ae514' but new value is '/Users/Joe/Projects/Repos/Thummer/alire/cache/dependencies/hal_0.3.0_095ae514'\n"

I didn't expect running the command would have side effects. I was under the impression it merely prints out its environment variables.  What does this error mean?

**Comment by mosteo** [comment]
It doesn't have side effects (any new exports are applied during the run only), but it seems `alr` is trying to set a variable to a different value than the one you had set before opening GNAT Studio.

So, how did you open it? `alr edit`? Are you changing to a different crate folder when running `alr` in the terminal pane? In principle the values should match and no error would be produced.

**Reply by Joebeazelman** [reply]
Thanks for the reply. 

I'm writing a GnatStudio plugin which enables seamless interoperation between GnatStudio and Alire without requiring the invocation of Alire's edit command from the system terminal. It's too easy to forget to do this, especially Mac and Windows. Moreover, there are times when you need to switch from one IDE to another and frequently reconfiguring the Alire's edit command gets to be cumbersome.

The plugin works by listening to a **project_changed** event and configures the environment variables based on the values obtained from Alire's **printenv** command. This is done before the project is loaded. In order to execute Alire, I added a field to the **Preferences** for specifying its path. Ideally,it should use the **$PATH** variable, but its unreliable. So far, result works perfectly, except for the error received when changing projects.

I investigated the cause by executing the Alire command under GnatStudio's terminal and received the same error. I'm baffled as to why it doesn't occur when it's executed outside GnatStudio in the OS terminal.

Despite this hiccup, the implementation has has gone smoothly compared to VS Code which lacks facilities for configuring its internal environment before the Ada Language Server Extension is launched. It's likely the solution requires a deeper extension based integration.

**Reply by mosteo** [reply]
> I investigated the cause by executing the Alire command under GnatStudio's terminal and received the same error. I'm baffled as to why it doesn't occur when it's executed outside GnatStudio in the OS terminal.

Somehow it seems that the variables being set in the environment are sticking around, so then the error is understandable. If you unset the vars on project change before calling `alr` it should work.

**Reply by Joebeazelman** [reply]
I figured it out. Changing some of the environment variables, specifically the non-Alire ones, doesn't return an error. 

Anyway, in case you haven't seen the announcement on the Alire Forum, I released the beta of my GNAT Studio plugin. It works nearly flawlessly, except for a minor issue where GNAT Studio doesn't broadcast a change project event when it launches with a selected Alire project. Consequently, the plugin has to force a reload in order for the IDE to recognize the Alire settings. This creates a slightly annoying pause and flash. If you have a solution for this, I'm interested.

**Reply by mosteo** [reply]
Nice to hear about the beta release! Unfortunately I'm not familiar with GNAT Studio's plugin system so I cannot help with that issue.
